Change ColOut to work with segmentation tasks

Landanjs opened the pull request:
A quick fix to get colout to work with segmentation, but maybe there should be a more general approach in the future? This would take more time though.

The batch version was pretty quick to setup: add a target argument to colout function that defaults to none, apply colout to target if it is present and matches the spatial dimensions of the input.

To get the sample version of colout, I converted the ADE20k dataset to a VisionDataset, included all transforms into a single object, and changed the colout transform to accept a 2-tuple of inputs. 

Wandb runs with 3 seeds can be found [here](https://wandb.ai/mosaic-ml/landan-deeplabv3-optimized?workspace=user-landanjs). Specific seeds: [sample method run](https://wandb.ai/mosaic-ml/landan-deeplabv3-optimized/runs/1eyzeaqq?workspace=user-landanjs) and [batch method run](https://wandb.ai/mosaic-ml/landan-deeplabv3-optimized/runs/33n14w05?workspace=user-landanjs)

coryMosaicML commented on composer/algorithms/colout/colout.py:
```diff
@@ -173,16 +201,17 @@ def _apply_sample(self, state: State) -> None:
-        new_inputs = colout_batch(inputs, p_row=self.p_row, p_col=self.p_col)
+        inputs, targets = state.batch_pair
+        assert isinstance(inputs, Tensor) and isinstance(targets, Tensor), \
+            "Multiple tensors not supported for this method yet."
```
This could be confusing as the method does support two tensors now. Maybe split this into separate checks for inputs and targets, or reword to specify that both inputs and targets should be single tensors?

coryMosaicML commented on composer/datasets/utils.py:
```diff
@@ -140,20 +143,35 @@ def add_vision_dataset_transform(dataset: VisionDataset, transform: Callable, is
+
+        if is_target_transformed:
+            assert dataset.transforms is not None
+            dataset.transforms.transforms.insert(insertion_index, transform)
```
The difference between the two cases here is hard to spot. Maybe a comment could point it out.

Landanjs replied:
Deleted this feature, so marking as resolved

coryMosaicML reviewed (commented):
Overall looks good! Main suggestions: have the resizing targets behavior be set via an algorithm hyper parameter rather than inferred from the shape of the targets, and possibly separate the dataset transform that operates on a pair vs. the one that operates on input only.

Landanjs commented on composer/algorithms/colout/colout.py:
```diff
@@ -73,6 +77,23 @@ def colout_batch(input: ImgT, p_row: float = 0.15, p_col: float = 0.15) -> ImgT:
+            Y_colout = Y_tensor[..., kept_row_idx, :]
+            Y_colout = Y_colout[..., :, kept_col_idx]
+            if not isinstance(target, torch.Tensor) or (target.ndim < Y_colout.ndim):
+                Y_colout = Y_colout.reshape(Y_colout.shape[-3:])
```
I stole this from the X_colout section. Was the X_colout code addressing a particular edge case? I can't tell. Here is the comment from X_colout:

```
# convert back to same type as input, and strip added batch dim if needed;
# we can't just reshape to input shape because we've reduced the spatial size
```

hanlint commented:
@Landanjs is this stale?

Landanjs commented on composer/cli/launcher.py:
```diff
@@ -317,7 +317,7 @@ def _print_process_exit_status(global_rank: int, process: subprocess.Popen):
         stderr=stderr,
     )
-    error_msg = [f"Global rank {global_rank} (PID {process.pid}) excited with code {process.returncode}"]
+    error_msg = [f"Global rank {global_rank} (PID {process.pid}) exited with code {process.returncode}"]
```
This should be exited, right? 😂

Landanjs commented:
### Major changes:
1. Removed features to add segmentation ColOut transform to `VisionDataset` since our streaming dataloader won't inherit from `VisionDataset` anyway. We need a higher level discussion on how to add individual transforms to datasets.
2. Attempted to add a `resize_target='auto'` option to capture both usability and flexibility.
3. Changed `colout_batch` to take a single input of either a single object or a 2-tuple as opposed to individual `input` and `target` arguments. This was inspired by the realization that we would like input and output types to match. With the change, tuples should return tuples and single objects should return single objects.
4. Added unit tests for situations where ColOut should also resize the target

I've decided to keep the ColOut transform to operate on both a single image or a 2-tuple of images. Although we can't use this feature right now, might be useful to others and is futureproofing for when we can use this feature. Maybe this is the wrong call? 😅 

### Things missing:
1. Update method card
2. Update docstring tests -> how do these even work?

Someone tell me I need to do this so I do it 😂 

### Results

[DeepLabv3+ADE20k ColOut](https://wandb.ai/mosaic-ml/landan-deeplabv3plus-streaming/groups/test_colout/workspace?workspace=user-landanjs) (in progress)
- **Question**: Throughput is about 1.22x, is this expected for 0.15 drop probability?

[ResNet50+ImageNet1k ColOut](https://wandb.ai/mosaic-ml/landan-debug/groups/test_resnet50_colout/workspace?workspace=user-landanjs) (in progress)
- **Question**: Throughput is about 1.3x, is this expected for 0.15 drop probability?

coryMosaicML commented on composer/algorithms/colout/colout.py:
```diff
@@ -39,16 +69,25 @@ def colout_batch(input: ImgT, p_row: float = 0.15, p_col: float = 0.15) -> ImgT:
+        p_row (float, optional): Fraction of rows to drop (drop along H). Default: ``0.15``.
+        p_col (float, optional): Fraction of columns to drop (drop along W). Default: ``0.15``.
+        resize_target (bool | str, optional): If ``sample`` is a tuple, whether to resize both objects in the tuple.
+            If set to 'auto', resizing both objects will be based on if the objects hav the same spatial dimensions.
```
Nit: "hav" -> "have"

Or maybe, "If set to 'auto', both objects will be resized if they have the same spatial dimensions."

Landanjs replied:
Ooops. Done!

coryMosaicML reviewed (approved):
Looks good to me! A few minor comments.

Landanjs commented on composer/algorithms/colout/colout.py:
```diff
@@ -26,9 +27,38 @@
-def colout_batch(input: ImgT, p_row: float = 0.15, p_col: float = 0.15) -> ImgT:
-    """Applies ColOut augmentation to a batch of images, dropping the same random rows and columns from all images in a
-    batch.
+def _should_resize_target(sample: Union[ImgT, Tuple[ImgT, ImgT]], resize_target: Union[bool, str]) -> bool:
```
Done!

Landanjs commented on composer/algorithms/colout/colout.py:
```diff
@@ -73,12 +112,29 @@ def colout_batch(input: ImgT, p_row: float = 0.15, p_col: float = 0.15) -> ImgT:
 class ColOutTransform:
     """Torchvision-like transform for performing the ColOut augmentation, where random rows and columns are dropped from
-    a single image.
+        up to two images.
```
What about explicitly stating "... up to two torch tensors or two PIL images."?
